packages/lending/src/adapter.ts

```
@@ -185,6 +185,299 @@ export class Adapter extends common.Web3Toolkit {
+      logics: [],
+    };
+    try {
+      // 1. ---------- validate params ----------
+      if (collateralAmount === '0') throw new OperationError('collateralAmount', 'ZERO_AMOUNT');
+
+      const srcCollateral = portfolio.findSupply(collateralToken);
+      if (!srcCollateral) throw new OperationError('collateralToken', 'UNSUPPORTED_TOKEN');
+
+      const destBorrow = portfolio.findBorrow(debtToken);
+      if (!destBorrow) throw new OperationError('debtToken', 'UNSUPPORTED_TOKEN');
+
+      const supplyAmountDelta = new common.TokenAmount(collateralToken.wrapped, collateralAmount).sub(
+        srcCollateral.balance
+      );
+      if (!srcCollateral.validateSupplyCap(supplyAmountDelta.amount)) {
+        throw new OperationError('collateralAmount', 'SUPPLY_CAP_EXCEEDED');
+      }
+
+      // 2. ---------- zap swap ----------
+      let supplyInput = new common.TokenAmount(collateralToken.wrapped, '0');
+      if (zapAmount !== '0' && !zapToken.wrapped.is(collateralToken.wrapped)) {
+        const swapper = this.findSwapper([zapToken, collateralToken]);
+        let swapQuotation: SwapperQuoteFields;
+        try {
+          swapQuotation = await swapper.quote({
+            input: new common.TokenAmount(zapToken, zapAmount),
+            tokenOut: collateralToken.wrapped,
+            slippage,
+          });
+        } catch {
+          throw new OperationError('zapAmount', 'NO_ROUTE_FOUND_OR_PRICE_IMPACT_TOO_HIGH');
+        }
+        const swapTokenLogic = swapper.newSwapTokenLogic(swapQuotation);
+        output.logics.push(swapTokenLogic);
+        // 2-1. the supply amount is the swap quotation output
+        supplyInput = swapQuotation.output;
+        output.afterPortfolio.supply(supplyInput.token, supplyInput.amount);
+      } else if (zapToken.wrapped.is(collateralToken.wrapped)) {
+        supplyInput = new common.TokenAmount(zapToken, zapAmount);
+        output.afterPortfolio.supply(supplyInput.token, supplyInput.amount);
+      }
```

Users calling the `openByCollateral` function should ensure they provide the correct `zapToken` and `zapAmount` . I believe it's unnecessary for us to handle cases where `zapAmount` equals "0", they can use `Leverage` instead.

If it is deemed necessary, we can include a check in this section by verifying that `zapAmount !== '0'` before proceeding further
```
if (zapAmount !== '0'){
  if (zapToken.wrapped.is(collateralToken.wrapped)) {
    // ...
  }else{
    // zapSupply
  }
 ```
 
   if not, we can throw the error on the validate-params part

---

I believe that both `collateralAmount` and `zapAmount` being set to 0 should be considered valid inputs and should return the default `OperationOutput`. When users enter 0 in the FE, we should not trigger an error popup. Please refer to `collateralSwap`, which checks `srcAmount` at the beginning.

---

IMO, `zapAmount` could be 0 but `collateralAmount` should be gte the user's supplied balance.